tests/utilities/test_auto_restart.py

```diff
@@ -1060,3 +1062,140 @@ def run(should_fail, resume):
     torch.testing.assert_allclose(total_train_batches, pre_fail_train_batches + post_fail_train_batches)
     for k in total_val_batches:
         torch.testing.assert_allclose(total_val_batches[k], pre_fail_val_batches[k] + post_fail_val_batches[k])
+
+
+class TestAutoRestartModelUnderSignal(BoringModel):
+    def __init__(self, should_signal: bool, failure_on_step: bool, failure_on_training: bool, on_last_batch: bool):
+        super().__init__()
+        self.should_signal = should_signal
+        self.failure_on_step = failure_on_step
+        self.failure_on_training = failure_on_training
+        self.on_last_batch = on_last_batch
+        self.seen_train_batches = []
+
+    def _signal(self, message):
+        if self.should_signal:
+            os.kill(os.getpid(), signal.SIGUSR1)
+            # small optimization to skip as soon as async signal has being triggered.
+            t0 = time()
+            while not self.trainer._terminate_gracefully:
+                t1 = time()
+                if (t1 - t0) > 0.1:
+                    self.trainer._terminate_gracefully = True
```

Comment:
why does this not use sleep?

Reply:
I am using a while loop to make sure I exit as fast as possible. This helped saved couple of seconds from the tests compared to a sleep.

Reply:
I had a look after not believing it and that happens because `self.trainer._terminate_gracefully` becomes `True` right after the signal is sent so the while loop never runs. This can be entirely removed

Reply:
This doesn't seem to happen for all the environment in the CI. If you remove it, it would fail.